src/Model/common.hpp

```diff
@@ -173,6 +177,58 @@ void setBlocks(T QgodLocal, Tmatrix S, Tarray1 traction_indices, Tarray2 velocit
     }
 }
 
+template <typename Tmaterial>
+seissol::eigenvalues::Eigenpair<std::complex<double>, NUMBER_OF_QUANTITIES> seissol::model::getEigenDecomposition (Tmaterial const& material, double zeroThreshold) {
+  std::array<std::complex<double>, NUMBER_OF_QUANTITIES*NUMBER_OF_QUANTITIES> AT;
+  auto ATView = yateto::DenseTensorView<2,std::complex<double>>(AT.data(), {NUMBER_OF_QUANTITIES, NUMBER_OF_QUANTITIES});
+  getTransposedCoefficientMatrix(material, 0, ATView);
+  std::array<std::complex<double>, NUMBER_OF_QUANTITIES*NUMBER_OF_QUANTITIES> A;
+  //transpose AT to get A
+  for (int i = 0; i < NUMBER_OF_QUANTITIES; i++) {
+    for (int j = 0; j < NUMBER_OF_QUANTITIES; j++) {
+      A[i+NUMBER_OF_QUANTITIES*j] = AT[NUMBER_OF_QUANTITIES*i+j];
+    }
+  }
+  seissol::eigenvalues::Eigenpair<std::complex<double>, NUMBER_OF_QUANTITIES> eigenpair;
+  seissol::eigenvalues::computeEigenvaluesWithEigen3(A, eigenpair);
+#ifndef NDEBUG
+  using CMatrix = Eigen::Matrix<std::complex<double>, NUMBER_OF_QUANTITIES, NUMBER_OF_QUANTITIES>;
+  using CVector = Eigen::Matrix<std::complex<double>, NUMBER_OF_QUANTITIES, 1>;
+  CMatrix eigenvectors = CMatrix(eigenpair.vectors.data());
+  CVector eigenvalues = CVector(eigenpair.values.data());
+  //check number of eigenvalues
+  //also check that the imaginary parts are zero
+  int evNeg = 0;
+  int evPos = 0;
+  for (int i = 0; i < NUMBER_OF_QUANTITIES; ++i) {
+    assert(std::abs(eigenvalues(i).imag()) < zeroThreshold);
+    if (eigenvalues(i).real() < -zeroThreshold) {
+      ++evNeg;
+    } else if (eigenvalues(i).real() > zeroThreshold) {
+      ++evPos;
+    }
+  }
+  assert(evNeg == 3);
```

Comment:
Is it good to hard-code that here?

Reply:
IMHO yes. After all it's only executed in debug mode and might help, if a scenario fails.